nautobot/dcim/tests/test_forms.py

```diff
@@ -189,3 +191,51 @@ def test_interface_label_count_mismatch(self):
 
         self.assertFalse(form.is_valid())
         self.assertIn("label_pattern", form.errors)
+
+
+class RackTestCase(TestCase):
+    def test_update_rack_site(self):
+        """Asset updating duplicate device caused by update to rack site is caught by rack clean"""
+        # Rack post save signal tries updating rack devices if site is changed, which may result in an Exception
+        # Asset this error is caught by RackForm.clean
+        tenant = Tenant.objects.first()
+        locations = Location.objects.filter(location_type=LocationType.objects.get(name="Campus"))[:2]
+        manufacturer = Manufacturer.objects.create(name="Manufacturer 1", slug="manufacturer-1")
+        devicetype = DeviceType.objects.create(model="Device Type 1", slug="device-type-1", manufacturer=manufacturer)
+        devicerole = Role.objects.get_for_model(Device).first()
+        status = Status.objects.get(name="Active")
+        racks = (
+            Rack.objects.create(name="Rack 1", site=locations[0], status=status),
+            Rack.objects.create(name="Rack 2", site=locations[1], status=status),
+        )
+
+        Device.objects.create(
+            name="device1",
+            device_role=devicerole,
+            device_type=devicetype,
+            site=racks[0].site,
+            rack=racks[0],
+            tenant=tenant,
+            status=status,
+        )
+        Device.objects.create(
+            name="device1",
+            device_role=devicerole,
+            device_type=devicetype,
+            site=racks[1].site,
+            rack=racks[1],
+            tenant=tenant,
+            status=status,
+        )
+        data = {
+            "name": racks[0].name,
+            "site": racks[1].site.pk,
+            "status": racks[0].status.pk,
+            "u_height": 48,
+            "width": RackWidthChoices.WIDTH_19IN,
+        }
+        form = RackForm(data=data, instance=racks[0])
+        self.assertEqual(
+            str(form.errors.as_data()["site"][0]),
+            str(["Device with `name` in ['device1'] and site=Site-2 already exists."]),
+        )
```

Comment:
```suggestion
```

Reply:
I think this test case is what stumped me in general and why I probably merged more in than I should have:

Isn't the issue still present when changing a Rack's location?

Reply:
```suggestion


class RackTestCase(TestCase):
    def test_update_rack_location(self):
        """Asset updating duplicate device caused by update to rack location is caught by rack clean"""
        # Rack post save signal tries updating rack devices if location is changed, which may result in an Exception
        # Asset this error is caught by RackForm.clean
        tenant = Tenant.objects.first()
        locations = Location.objects.filter(location_type=LocationType.objects.get(name="Campus"))[:2]
        manufacturer = Manufacturer.objects.create(name="Manufacturer 1", slug="manufacturer-1")
        devicetype = DeviceType.objects.create(model="Device Type 1", slug="device-type-1", manufacturer=manufacturer)
        devicerole = Role.objects.get_for_model(Device).first()
        status = Status.objects.get(name="Active")
        racks = (
            Rack.objects.create(name="Rack 1", location=locations[0], status=status),
            Rack.objects.create(name="Rack 2", location=locations[1], status=status),
        )

        Device.objects.create(
            name="device1",
            device_role=devicerole,
            device_type=devicetype,
            location=racks[0].location,
            rack=racks[0],
            tenant=tenant,
            status=status,
        )
        Device.objects.create(
            name="device1",
            device_role=devicerole,
            device_type=devicetype,
            location=racks[1].location,
            rack=racks[1],
            tenant=tenant,
            status=status,
        )
        data = {
            "name": racks[0].name,
            "location": racks[1].location.pk,
            "status": racks[0].status.pk,
            "u_height": 48,
            "width": RackWidthChoices.WIDTH_19IN,
        }
        form = RackForm(data=data, instance=racks[0])
        self.assertEqual(
            str(form.errors.as_data()["location"][0]),
            str([f"Device with `name` in ['device1'] and location={locations[1]} already exists."]),
        )
```